### PR TITLE
Metric getattrs

### LINKED ISSUE #414: Add helper ops for metrics arithmetics
Following the discussion from [here](https://github.com/pytorch/ignite/pull/410#issuecomment-457105415)

> I was wondering whether it is possible to introduce a metric that performs mean, Mean, and compact the following piece of code:

```python
    precision = Precision(average=False)
    recall = Recall(average=False)
    F1 = precision * recall * 2 / (precision + recall + 1e-20)
    F1 = MetricsLambda(lambda t: torch.mean(t).item(), F1)
```
> to:
```python
    precision = Precision(average=False)
    recall = Recall(average=False)
    metrics = Mean(precision * recall * 2 / (precision + recall + 1e-20))
```

Idea is to provide some helper classes like `Mean`, `Sum` and others.

cc @amitibo

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes #414

Description:
As discussed in https://github.com/pytorch/ignite/pull/417#issuecomment-461652286

Check list:
* [x] New tests are added (if a new feature is modified)
* [ ] New doc strings: text and/or example code are in RST format
* [x] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/metrics/test_functional.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        assert set(state.metrics.keys()) == set([metric_name, ])
+        np_y_pred = np.argmax(y_pred.numpy(), axis=-1).ravel()
+        np_y = y.numpy().ravel()
+        assert state.metrics[metric_name] == approx(compute_true_value_fn(np_y_pred, np_y))
+
+    precision_1 = Precision(average=False)
+    precision_2 = Precision(average=False)
+    mean_summed_precision1 = F.mean(precision_1 + precision_2).item()
+    mean_summed_precision2 = (precision_1 + precision_2).mean().item()
```
Comment (zasdfgbnm) [inline tests/ignite/metrics/test_functional.py]: @vfdev-5 Unfortunately the API here are slightly different from those on https://github.com/pytorch/ignite/pull/417

**Comment by zasdfgbnm** [comment]
@vfdev-5 How does this looks like? If the API looks good, then I will update docs.

**Comment by vfdev-5** [comment]
@zasdfgbnm thank you, really awesome code ! 

I'm getting the understanding of what it does and it opens the boundaries and not sure how could we use it in our applications but feels great :) 
For example, chain operations on tensors:
```python
m = (precision_1 + precision_2).pow(2.0).mean()
```
Thinking more about this, I have a slight impression that we "kill a fly with a bazooka" :) 
At the begining, we needed to reduce vectors into scalars and now we have any attribute of torch applied on metrics...

About `ignite.metrics.functional`, I think this is pure private stuff and we could keep inside `Metric` ?

Maybe we can provide some more magic and override somehow the wrapper to produce `.item()` at the end of the operation. Such that we provide needed reduction and do not make all this stuff too complex and error prone.

**Comment by zasdfgbnm** [comment]
@vfdev-5 I changed `Metric.completed` and now `item` is handled automatically. What do you mean keeping functional inside `Metric`? do you mean we just add `__getattr__` for `Metric` and discard the whole `functional.py`? I think we need this, because not all pytorch operators are methods, some of them you need to access it through `torch.xxx(tensor)` but not `tensor.xxx()`

**Comment by zasdfgbnm** [comment]
Seems that module __getattr__ was introduced in [PEP562(]https://www.python.org/dev/peps/pep-0562/), available only on python 3.7...

**Comment by zasdfgbnm** [comment]
I temporarily disabled the two failing RL examples from CI. We could reopen this after https://github.com/pytorch/ignite/issues/424 is resolved.

**Comment by zasdfgbnm** [comment]
@vfdev-5 I updated the document, looks good to myself now.

**Comment by vfdev-5** [comment]
@zasdfgbnm, I'll take a look at the code  in more details asap.

**Comment by amitibo** [comment]
@zasdfgbnm and @vfdev-5, I like this approach very much. I wonder if it enables passing arguments to the functions, .e.g `axis` etc.

**Comment by zasdfgbnm** [comment]
@amitibo It does

**Comment by vfdev-5** [comment]
@zasdfgbnm I checked out the code, it is really cool stuff you proposed here 👍  and I think we will keep the part added into `Metric`. Concerning `import ignite.metrics.functional as F` and `_pep562.py` it is probably too early to introduce that. As it just gives only an equivalent usage and for that we need to create a helper file,  so I propose to remove that and to keep the API as
```
precision = Precision(average=False)
recall = Recall(average=False)
F1 = (precision * recall * 2 / (precision + recall)).mean()
```
without `F.*`. 

What do you think, @zasdfgbnm  @amitibo ?

**Comment by zasdfgbnm** [comment]
@vfdev-5 If `import ignite.metrics.functional as F` was just an alias of metric methods, the I had no preference on having it or not. My only concern is that, there are some operators not accessible using method, for example `torch.meshgrid` has no equivalent `tensor.meshgrid` version. So functional would be a super set of methods.  But I don't know if there will be any user interested in torch operators that are inside functional but not a method.

Anyway, just let me know your decision. I will just follow.

**Comment by vfdev-5** [comment]
@zasdfgbnm that's true, functional part gives more available operators. 
OK, for this PR we can leave it with `Metric.__getattr__` only and if we see any interest into functional part we'll add it after. So, let's adapt the current PR that way.

**Comment by zasdfgbnm** [comment]
@vfdev-5 done

**Comment by vfdev-5** [comment]
@anmolsjoshi review and merge please
